internal/policies/mount/adsys_mount/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+
+    let error = errors
+        .iter()
+        .any(|MountError { error, .. }| !error.matches(gio::IOErrorEnum::AlreadyMounted));
+
+    if error {
+        return Err(AdsysMountError::MountError);
+    }
+
+    Ok(())
+}
+
+fn msg_handler(
+    msg: Msg,
+    errors: &Mutex<Vec<MountError>>,
+    main_loop: &glib::MainLoop,
+    mounts_left: &mut usize,
+) -> glib::Continue {
+    let Msg { path, status } = msg;
+    match status {
+        Ok(_) => debug!("Mounting of {} was successful", path),
+        Err(error) => {
+            warn!("Failed when mounting {}", path);
+            errors.lock().unwrap().push(MountError { path, error });
+        }
+    };
+    *mounts_left -= 1;
+
+    // Ends the main loop if there are no more mounts left.
+    if *mounts_left == 0 {
+        main_loop.quit();
+    }
+    glib::Continue(*mounts_left != 0)
+}
+
+/// Reads the file and parses the mount points listed in it.
+fn parse_entries(path: &str) -> Result<Vec<Entry>, std::io::Error> {
+    debug!("Parsing file {} content", path);
+
+    let mut parsed_entries: Vec<Entry> = Vec::new();
+
+    let content = fs::read_to_string(path)?;
+    for line in content.lines() {
+        if line.is_empty() {
+            continue;
+        }
+
+        parsed_entries.push(match line.strip_prefix("[anonymous]") {
+            Some(s) => Entry {
+                path: s.to_string(),
+                is_anonymous: true,
+            },
+            None => Entry {
+                path: line.to_string(),
+                is_anonymous: false,
+            },
+        });
+    }
+
+    Ok(parsed_entries)
+}
+
+/// Handles the mount operation to mount the specified entry.
+fn mount_entry(entry: Entry, tx: glib::Sender<Msg>) {
+    debug!("Mounting entry {}", entry.path);
+
+    let f = gio::File::for_uri(&entry.path);
+
+    let mount_op = gio::MountOperation::new();
+
+    if entry.is_anonymous {
+        debug!("Anonymous mount requested for {}", entry.path);
+        mount_op.set_anonymous(true);
+    }
+
+    mount_op.connect_ask_password(ask_password_cb);
+
+    // Callback invoked by gio after setting up the mount.
+    let mount_handled_cb = move |r: Result<(), glib::Error>| {
+        let msg = Msg {
+            path: entry.path,
+            status: r,
+        };
+
+        if let Err(e) = tx.send(msg) {
+            error!("Failed to send message in the channel: {}", e);
+        }
+    };
+
+    f.mount_enclosing_volume(
+        gio::MountMountFlags::NONE,
+        Some(&mount_op),
+        gio::Cancellable::NONE,
+        mount_handled_cb,
+    );
+}
+
+/// Callback that is invoked by gio when prompted for password.
+fn ask_password_cb(
+    mount_op: &gio::MountOperation,
+    _: &str,
+    _: &str,
+    _: &str,
+    flags: gio::AskPasswordFlags,
+) {
+    if mount_op.is_anonymous() && flags.contains(gio::AskPasswordFlags::ANONYMOUS_SUPPORTED) {
+        // Unsafe block is needed for data and set_data implementations in glib.
+        unsafe {
```

Comment:
I tend to follow this policy regarding unsafe blocks:

https://std-dev-guide.rust-lang.org/documentation/safety-comments.html

Usually, `unsafe fn foo()` will have a comment describing why it is unsafe, i.e. which invariants need to be upheld by the calling code. Here, `set_data()` and `data()` have indeed a "Safety" section: https://docs.rs/glib/latest/glib/object/trait.ObjectExt.html#tymethod.set_data

> This function doesn’t store type information

> The returned pointer can become invalid by a call to set_qdata, steal_qdata, set_data or steal_data.
>
> The caller is responsible for ensuring the returned value is of a suitable type

A key principle in Rust is local reasoning: you should be able to figure stuff out at the function level. This is true for type inference (all types can be deduced from either local variable declarations or deducted from the function signature), but also for safety reasoning.

Here, the only way to make sure the invariant is upheld (we only store booleans in there, and nowhere else is `set_data` called) is to grep through the whole codebase, which means that your function `ask_password_cb` should be marked `unsafe`. The place where you can enforce that invariant is actually when you connect the callback, as there you have a grasp over the `MountOperation` lifecycle, know that it's only used locally (and assume that `mount_enclosing_volume` don't do nasty stuff with `set_data`, but that should be documented as well).

So, in short, this is what I'm proposing:

```rust
/// Handles the mount operation to mount the specified entry.
fn mount_entry(entry: Entry, tx: glib::Sender<Msg>) {
    debug!("Mounting entry {}", entry.path);

    let f = gio::File::for_uri(&entry.path);

    let mount_op = gio::MountOperation::new();

    if entry.is_anonymous {
        debug!("Anonymous mount requested for {}", entry.path);
        mount_op.set_anonymous(true);
    }

    mount_op.connect_ask_password(|mount_op, _, _, _, flags| {
        if mount_op.is_anonymous() && flags.contains(gio::AskPasswordFlags::ANONYMOUS_SUPPORTED) {
            // SAFETY: no use of `set_data` and friends on the local mount_op, assuming
            // `mount_enclosing_volume` also doesn't.
            unsafe { ensure_single_anonymous_access(mount_op) };
            return;
        }
        // Checks if the machine has a kerberos ticket defined.
        if std::env::var("KRB5CCNAME").is_ok() {
            debug!("Kerberos ticket found on the machine.");
            mount_op.reply(gio::MountOperationResult::Handled);
            return;
        }

        warn!("Kerberos ticket not available on the machine.");
        mount_op.reply(gio::MountOperationResult::Aborted);
    });

    // Callback invoked by gio after setting up the mount.
    let mount_handled_cb = move |r: Result<(), glib::Error>| {
        let msg = Msg {
            path: entry.path,
            status: r,
        };

        if let Err(e) = tx.send(msg) {
            error!("Failed to send message in the channel: {}", e);
        }
    };

    f.mount_enclosing_volume(
        gio::MountMountFlags::NONE,
        Some(&mount_op),
        gio::Cancellable::NONE,
        mount_handled_cb,
    );
}

/// # Safety
///
/// This function assumes it has exclusive use of `mount_op`'s `set_data()` and related methods.
unsafe fn ensure_single_anonymous_access(mount_op: &gio::MountOperation) {
    // SAFETY: per function invariant
    if let Some(data) = unsafe { mount_op.data::<bool>("state") } {
        // Ensures that we only try anonymous access once.
        // SAFETY: per function invariant
        if unsafe { *data.as_ptr() } { // could also have been `if *(unsafe { data.as_ref() })` but that's rather ugly
            warn!("Anonymous access denied.");
            mount_op.reply(gio::MountOperationResult::Aborted);
        }
    } else {
        debug!("Anonymous is supported by the provider.");
        // SAFETY: per function invariant
        unsafe { mount_op.set_data("state", true) };
        mount_op.reply(gio::MountOperationResult::Handled);
    }
}
```

Note that in Rust < 1.65 this will give you warnings, as the community practices around `unsafe` are still evolving. Previously, being in a `unsafe fn` not only meant that the function couldn't be called without `unsafe` but also that its contents could use unsafe operations without further `unsafe` blocks, and so an old lint considered them redundant. To disable these old warnings (and actually warn if *not* using an `unsafe` block), add `#[warn(unsafe_op_in_unsafe_fn)]` at the top of `lib.rs`.